fairlearn/metrics/_metric_frame.py

```diff
@@ -0,0 +1,552 @@
+# Copyright (c) Microsoft Corporation and Fairlearn contributors.
+# Licensed under the MIT License.
+
+import copy
+import logging
+import numpy as np
+import pandas as pd
+from typing import Any, Callable, Dict, List, Optional, Union
+from sklearn.utils import check_consistent_length
+
+from fairlearn.metrics._input_manipulations import _convert_to_ndarray_and_squeeze
+from ._function_container import FunctionContainer, _SAMPLE_PARAMS_NOT_DICT
+from ._group_feature import GroupFeature
+
+logger = logging.getLogger(__name__)
+
+_SUBGROUP_COUNT_WARNING_THRESHOLD = 20
+
+_BAD_FEATURE_LENGTH = "Received a feature of length {0} when length {1} was expected"
+_SUBGROUP_COUNT_WARNING = "Found {0} subgroups. Evaluation may be slow"
+_FEATURE_LIST_NONSCALAR = "Feature lists must be of scalar types"
+_FEATURE_DF_COLUMN_BAD_NAME = "DataFrame column names must be strings. Name '{0}' is of type {1}"
+_DUPLICATE_FEATURE_NAME = "Detected duplicate feature name: '{0}'"
+_TOO_MANY_FEATURE_DIMS = "Feature array has too many dimensions"
+_SAMPLE_PARAM_KEYS_NOT_IN_FUNC_DICT = \
+    "Keys in 'sample_params' do not match those in 'metric'"
+
+
+class MetricFrame:
+    """Contains a collection of group fairness metrics.
+
+    This data structure stores and manipulates disaggregated values for any number of underlying
+    metrics. At least one sensitive feature must be supplied, which is used
+    to split the data into subgroups. The underlying metric(s) is(are) calculated
+    across the entire dataset (made available by the :attr:`.overall` property) and
+    for each identified subgroup (made available by the :attr:`.by_group` property).
+
+    The only limitations placed on the metric functions are that:
+
+    * The first two arguments they take must be ``y_true`` and ``y_pred`` arrays
+    * Any other arguments must be *sample* based, and will be split up along with
+      the ``y_true`` and ``y_pred`` arrays
+
+    The interpretation of the ``y_true`` and ``y_pred`` arrays is up to the
+    underlying metric - it is perfectly possible to pass in lists of class
+    probability tuples. The return value of the metric is also very flexible;
+    there is no problem around returning confusion matrices.
+
+    Once calculated, scalar group fairness metrics can be aggregated over
```

Comment:
->
Group fairness metrics are obtained by methods that implement
various aggregators over group-level metrics, such such as the
maximum, minimum, or the worst-case difference or ratio.

Reply:
Updated.